[Team3] Report (#16)
SOMAS2021/SOMAS-Final-Report: 005_team_3_agent_design/005_team_3_agent_design.tex

\chapter{Team 3 Agent Design}\label{team_3_agent_design}

\section{The Agent}\label{the_agent}

> The agent is designed to show different combinations of behaviour depending on three variables we have used to define any agent: stubbornness, \cite{chun_2005},  morality, \cite{waller_2015}, and mood. These three variables, defined in a range of 0 to 100 will change as the agent encounters a series of scenarios, with its personality shifting as events unravel.  The value of the different variables will affect how the agent takes decisions and acts, all of which will be explained in the following pages. \par

Reviewer: Remove the commas before citations

%%ADD THE MAIN CODE!!!

> As the main function from which all the functionality of the agent is bases we have the \texttt{func (a *CustomAgent3) Run()}. The agent follows then next steps:

Reviewer 2: The reader doesn't care about the specifics of `a.Run()`

You could make things a lot simpler by just saying "each day, the agent performs the following actions"

Author: Done

\begin{enumerate}
    \item The agent starts by updating its personality variables at the beginning of each day calling the necessary function. 
    \item If the agent was just ‘born’ it will learn an initial value for the average food consumed. 
    \item In the case the agent has changed floors the personality variables are changed depending on the characteristics of the change. 
    \item In the case the platform is not empty the agent will eat, the quantity depending on \texttt{a.takeFoodCalculation()}.
    \item Finally the agent will read and send any messages it considers.
\end{enumerate}

> \subsection{Agent knowledge}

Reviewer 2: It's worth capitalising all of the headers

Author: Done

Agent 3 will be able to remember facts in order to make decisions. The variables in their memory are:
\begin{enumerate}
    \item \texttt{floors []int}: stores the floors the agent has been in before. This information is used on reshuffles to set our mood. 
    \item \texttt{lastHp int}: stores the last recorded HP value. Currently used to check a new day has started.
    \item \texttt{friends map[uuid.UUID]float64}: the agent is aware of the people they have met during their time in the tower. We assigned the agents uuid with our affection for them. Affection range is 0-1 (0 = dislike, 1 = like). 

>     \item \texttt{foodLastEaten food.FoodType}: stores food last eaten.

Reviewer 2: I wouldn't namespace in a report - int conveys enough information here (as a reader won't always have a codebase)

Reviewer 2: (someone external to this project won't know what food.FoodType means)

Author: The text has changed, not applicable anymore.

    \item \texttt{foodLastSeen food.FoodType}: stores food last seen in the platform
    \item \texttt{foodMovingAvg float64}: stores moving average of food consumed
    \item \texttt{agentAge int}: stores how old (in days) the agent is.

>     \item \texttt{treatyProposed messages.Treaty}: stores whether we already have a treaty with the person above

Reviewer 2: same deal, I'd call "Treaty" and reference this to the main text, as a reader doesn't care which file it's imported from

Author: The text has changed, not applicable anymore.

\subsection{Agent generation}
%%TODO: have to add the code snippet
The function generates a new agent by initializing its variables, knowledge and decisions. All knowledge and decisions are defined initially as empty except for our last HP, which is initialized as 100 (as all agents are). \par
The variables of our agent, which define the personality it has, are randomly allocated to generate different agents. All three variables are defined to be in the range 0 to 100. Therefore morality and mood are randomly allocated to a value between 0 and 100. To ensure our agent is not completely deaf from the start we limit the initial stubbornness to 75. This does not limit the agent once it starts interacting and taking decisions. Which means it could possibly end up with a stubbornness value of 100 if the situations it goes through are auspicious for it.

\subsection{Agent Strategy}

> Our agent follows the following strategy depending on the three agent variables:

Reviewer 2: I wouldn't use "our" in an academic report. Use "this agent" or "it" contextually. You have a lot of uses of "our agent" and sometimes "the agent," (which would be correct). Best not to mix them up

Author: Done

\begin{enumerate}
    \item When morality is high, agents are more willing to listen and obey other agents.
    \item When morality is low, agents are less willing to listen and choose to eat more than they would when they have high morality.
    \item When stubbornness is high, agents will be “deaf” to most messages and not react to them.
    \item When stubbornness is high, agents are more willing to listen to messages.
    \item When mood is high, the actions taken are beneficial for more people.
    \item When mood is low, the actions taken are more beneficial for ourselves.
\end{enumerate}

\section{Agents emotions}\label{agents_emotions}

> The following functions act as helpers for the rest of the agent code, enabling changes in the agents variables and reactions due to the value of these variables, as well as reactions to friendships. 

Reviewer 2: ```suggestion
The following functions act as helpers for the rest of the agent code, enabling changes in the agent's variables and reactions due to the value of these variables, as well as reactions to friendships. 
```

Reviewer 2: agents' also works if you're referring to multiple

Author: Done

\subsection{Read function}

> \texttt{func (a *CustomAgent3) read() bool}: our agent's stubbornness will limit the amount of messages it might read and react to. Each time we receive a message we call the \texttt{read()} function. This function will take into account the current stubbornness of the agent and return if the message must be read, and reacted to, or just discarded. 

Reviewer 2: ```suggestion
\texttt{func (a *CustomAgent3) read() bool}: our agent's stubbornness will limit the amount of messages it might read and react to. Each time an agent receives a message we call the \texttt{read()} function. This function will take into account the current stubbornness of the agent and return if the message must be read, and reacted to, or just discarded. 
```

"We" should refer to the programmer (you guys!), not the agent :)

This function works by randomly generating a number between 0 and 100. If this number is higher to the current stubbornness the message is read, if not it will be discarded. 

\subsection{Friendship Function}
\begin{sloppypar} %%Gives style warning so this sloppypar solves it

> \texttt{func (a *CustomAgent3) updateFriendship(friend uuid.UUID, change int)}: changes the value of the friendships using Q-Learning. Change is either -1 or 1. \par

Reviewer 2: similar to earlier comments, uuid.UUID conveys no meaning that UUID doesn't

The reader doesn't need to see the package

\chapter{Team 3 Agent Design}\label{team_3_agent_design}

\section{The Agent}\label{sec:the_agent}

> Team 3’s agents are based on the Theory of Reasoned Action (TRA). The agents are designed to have a shifting personality dependent on the three variables that this theory links to the behaviours of an agent. These three variables, together with a social network, define their personality as a whole, \cite{victor} and influence the actions they will take.\par

Reviewer: Maybe start with Team 3's agent design is based. Remove comma before citation

\begin{figure}[htb]
    \centering
    \includegraphics[width=0.5\linewidth]{005_team_3_agent_design/images/image1.jpg}
    \label{fig:TRA-diagram}
\end{figure}

> The Theory of Reasoned Action was developed by Martin Fishbein in 1967. The TRA states that the behavioural intentions of individuals can be attributed to three interdependent factors: Perceived behaviour control, the subjective norm, and the attitude of the individual, as you can see in the figure above, \cite{norazalan}. Although it was recognised from the start, it must be noted that the TRA is unable to accurately and consistently predict non-volitional behaviours of the individual. Therefore, TRA can only be applied to those actions in which a reason can be found, actions that occur as a reaction to others and that the individual has control over, \cite{TRA}. This is applicable to our scenario, in which the agent must make a decision on how much food to eat after deliberating on their social network’s messages. \par

Reviewer: Remove comma before citation and you should reference the figure rather than refer to figure above.

> The application of TRA, in this case, is ideal given that reversing the methodology enables us to define an individual through the 3 core personality traits. The theory is a valid methodology to predict future behaviour by observing occurring actions and finding those components that caused them. Those components of behaviour can also be extrapolated to define an individual's behaviour and used to define it from the start. Therefore building the main characteristics of the personality of the individual, although lacking the uncontrollable or instinctive reactions. \par

Reviewer: change "and" to and are used

Author: If your correction is for 'an individual's behaviour and used to define it from the start' I don't find it appropriate

> The application of TRA, in this case, is ideal given that reversing the methodology enables us to define an individual through the 3 core personality traits. The theory is a valid methodology to predict future behaviour by observing occurring actions and finding those components that caused them. Those components of behaviour can also be extrapolated to define an individual's behaviour and used to define it from the start. Therefore building the main characteristics of the personality of the individual, although lacking the uncontrollable or instinctive reactions. \par

Reviewer: Change Therefore to Hence

Following the Theory of Reasoned Action, we have defined the agents’ emotion variables to be: 
\begin{enumerate}

>     \item Stubbornness: it addresses perceived behaviour control, which is the influence that external agents have on the decisions that are to be taken. It is applied by the agent to ignore or read incoming messages and succumb or not to group pressure. 

Reviewer: and succumb -> and to succumb

    \item Morality: it addresses the subjective norm, which is one's idea of what is socially perceived to be right or wrong.
    \item Mood: it defines the attitude towards the action to take.
\end{enumerate}

> These variables are modified as agents interact, which lead our agent to make different decisions during its time in the tower. All possible changes to these variables are done through ranges with the exact change each time being randomised, in other words inside a specified range any of the variables will change a different amount as a reaction for the same action. This course of action is based on the substantive uniqueness of each behaviour as defended by Martin Fishbein, \cite{TRA}. This characteristic of randomness in the agents’ behaviour fulfilled the role of covering the fact that TRA does not take into account non-volitional behaviours, making the agent act as much as possible in a human-like behaviour. \par

Reviewer: Again, no comma before citation

Reviewer 3: ```suggestion
These variables are modified as agents interact, which lead our agent to make different decisions during its time in the tower. All possible changes to these variables are done through ranges with the exact change each time being randomised, in other words inside a specified range any of the variables will change a different amount as a reaction for the same action. This course of action is based on the substantive uniqueness of each behaviour as defended by Martin Fishbein \cite{TRA}. This characteristic of randomness in the agents’ behaviour fulfilled the role of covering the fact that TRA does not take into account non-volitional behaviours, making the agent act as much as possible in a human-like behaviour. \par
```

> The next step after identifying our agent strategy was to think about the scenario and how the problem should be solved. Our first approach used classical game theory, which states that under normal circumstances, an unbiased agent would always prioritise themselves and their needs over anything else. Hence, this dictates that each agent should not act in a moral way at all. In our scenario, this leads to agents taking selfish amounts of food, causing agents below to have less food and potentially die. However, we found research demonstrating how humans in a social environment tend to make decisions benefiting the collective, \cite{batson_batson_todd_brummett_shaw_aldeguer_1995}. This contradicted traditional game theory views and made this approach unreasonable when trying to recreate a human type behaviour, following the two most common explanations of why humans violate classic game theory: enlightened self-interest and group identity. 

Reviewer: No comma before citation

\subsection{Action Process}

> The Agent follows the following steps during their day: 

Reviewer 4: ```suggestion
The Agent takes the following steps during their day: 
```

\begin{enumerate}

>     \item The agent starts by updating their defining variables at the beginning of each day based on their HP. This acts as the agents checking in on themselves and changing their attitude(mood) and morality depending on their condition(HP).

Reviewer: space before brackets

Reviewer 3: ```suggestion
    \item The agent starts by updating their defining variables at the beginning of each day based on their HP. This acts as the agents checking in on themselves and changing their attitude(mood) and morality depending on their condition (HP).
```

    \item The amount of food last eaten.
    \item The amount of food last seen on the platform.
    \item A moving average of food consumed.
    \item The age of the agent.
    \item The treaty we have proposed until it is accepted or rejected.
    \item The estimation of the reshuffle.
    \item The HP our neighbour above last told us they had.
    \item The HP our neighbour below last told us they had.
\end{enumerate}

\subsection{Agent Generation}
The defining variables of our agents are randomised when they are generated in order to avoid preconditioned agents and mimic different personalities. All three variables are defined to be in the range 0 to 100. Morality and mood are randomly allocated to a value in the range, but stubbornness is limited to a number between 0 and 75. As explained in our agent parameters, stubbornness represents the effect of external influences on an agent. A stubbornness of 100 would mean this agent is completely unwilling to listen and would ignore all incoming messages, effectively making them an agent that disturbs the tower. In order for this to not be the case, we added the limitation to the value of stubbornness.\par

> As Shao et al. have shown in “Beyond Moral Reasoning: A Review of Moral Identity Research and Its Implications for Business Ethics”, \cite{shao_aquino_freeman_2008}, one’s moral action is most heavily influenced by their moral personality, as opposed to their moral reasoning, which empirically seems to show only a minor influence on the moral behaviour of an agent, \cite{blasi_1983}. Hence, we used this variance and unpredictability to initialise our agents randomly on the moral scale. This insight also allowed us to witness a purely emergent strategy and nature as opposed to a pre-programmed inherent morality.\par

Reviewer: Remove title and just cite. Remove commas before citations

> While the initial randomness allowed for interesting strategies to occur, we also wanted our agents to adapt their morality on the basis of their community and associations, as argued by Shao. Making it possible by having all actions an agent does affect its emotion variables. 

Reviewer: Cite Shao

\section{Agent Communication}\label{sec:agent_communication}

> Communication between agents is fundamental for the possibility of collaboration and self-organisation between agents. Crowd behaviour is considered to be generated by individuals, context-dependent and dynamic, as defended by the modern foundation of crowd behaviour, \cite{0d9bc1ee81234780b2bb6ecd02762d56}. In this project, communication between agents is based on two concepts, treaties and direct messages, which are equally exploited by our agent and combined with the conceptualization of friendship levels. 

Reviewer: Remove comma before citation

\subsection{The Agent’s Social Network}
The agent forms relationships with others surrounding them as they communicate with each other. The agent remembers everyone they have met and assigns a level of friendship to them depending on their interactions. By doing this, the agent forms a social network that will affect their decisions, something that the Theory of Reasoned Action explores in the perceived behaviour control aspect of the theory. It is shown that all three main variables are linked together and for this reason, the agent will have a better or worse opinion of someone they have just met depending on their morality at that point. \par

> Relationships are of high esteem for the agent as it follows the Aristotelian concept of friendship which defends friendship as “the finest way we exercise practical reasoning”. Quoting Aristotle: “Friendship seems to hold cities together, and lawgivers seem to care for it more than for justice … and when men are friends, they have no need for justice”. Friendship is considered the best form of human moral action by The Nicomachean Ethics, \cite{sokolowski}, and is fundamental in the decision-making process of the agent and its response to communication started by other agents. \par 

Reviewer: Remove comma before citation

> McCullough et al. investigate thoroughly the nature of prosocial action and gratitude’s place in transactions, \cite{doi:10.1111/j.1467-8721.2008.00590.x}. While this is difficult to realise fully in the code, we aimed to use the definitional guidelines set out, to relate our agent’s action and emotion to some of the more nuanced but socially beneficial emotions. Our friendship function allows us to assign a level of gratitude to our agent in response to the prosocial action of the other agents. In this way, we could more effectively alert the agent to the prosocial nature of another’s actions. By utilising the friendship function to, in turn, behave more prosocially to those who have done so to us, we embedded a system of gratitude in our agents’ actions. \par

Reviewer: Remove comma before citation

We graduated our response by negotiating the selflessness of the action of the foreign agent. That is to say, in our game, the more severe the sacrifice on the behalf of the foreign actor, the more grateful we became. This is based on McCullough’s definition that one experiences gratitude when “they have benefited from someone’s costly, intentional, voluntary effort on their behalf”. \par 

> Emotions allow us to respond to reality in a self-protective and self-enhancing way, \cite{ekman_1992}, it was this response we wanted our agent to have so that the emotions would govern their actions, by either leading them on a self-protective path, if circumstances such as selfish agents prevailed, or a self-enhancing manner, as cooperation would beget further cooperation through our system of gratitude.\par 

Reviewer: Remove comma before citation

> Emotions allow us to respond to reality in a self-protective and self-enhancing way, \cite{ekman_1992}, it was this response we wanted our agent to have so that the emotions would govern their actions, by either leading them on a self-protective path, if circumstances such as selfish agents prevailed, or a self-enhancing manner, as cooperation would beget further cooperation through our system of gratitude.\par 

Reviewer: Break up last sentence. It is too long.

    \item If the agent decides to read the message it is analysed and matched with the type of response or responses acceptable for the message. 
    \item The agent will note the agent that sent the message and consider the friendship level it shares with the sender. Friendship together with the morality level and the content of the message itself will decide the exact content of the response.
    \item The content of the message and its sender will affect the agents’ emotional variables and the friendship level with the sender. This change affects the future reaction to communication done by other agents and the decisions taken by the agent from that point onwards. 
\end{enumerate}
To ensure a minimum level of communication all messages read will be answered by the agent. 

\subsection{Reaction to Received Messages}
The changes are dependent on four criteria to determine initial agent characteristic manipulation. These four variables are the variables present upon the receipt of the message and somewhat representative of the type 1 brain system discussed in \cite{kahneman_tversky_2000}, \cite{frankish_2010}. \par

> The initial emotional response is based on the variables of the immediate past, the hp moving average on the last five ticks, opposed to the overall hp average, or welfare average. This is to represent the subconscious layer of emotional response, as seen in, \cite{kamble_2021}, whereby our response to the situation is more irrational and illogical than the conscious mind perceives. \par

Reviewer: Remove comma before citation

Summing the net total of all the initial responses to zero, allows us to explore the oscillatory nature of emotion while avoiding the possible pitfall of any dominant emotions. \par

> \begin{table}[htb]

Reviewer: Inconsistent capitalisation

Reviewer: This is for all tables below

    \end{tabular}
\end{table}
This zero-sum agent is how our agent is currently set, it represents the neutral agent. One who possesses no inherent emotional bias. This was done by balancing the positive and negative reactions to a friend, and those to a non-friend separately. We noticed the need for this when our agent would initially get stuck at the +ve and -ve limits of certain emotions. As future work, it would be exciting to see the emergent strategies which we hope to observe as we introduce biases. For instance, an agent whose net sum of all interactions leaves them with a stubbornness increase, coupled with an inherently moral agent, may create strategies yet thought of. 

\subsection{Proposing a Treaty}
When the agent HP is critical, they will ask for help by either asking for the person upstairs to leave at least the amount of food necessary to survive or, if it doesn’t have any active treaties, a treaty that states: if HP is higher than 20 leave 95\% of the food in the platform for 3 days. At this time the agent's self-preservation instinct takes precedence and their objective becomes to eat enough food to get out of the critical state before the agent dies.

> When the agent is not in critical condition, it will be able to propose other treaties depending on their morality. The first treaty is detrimental for the tower and the other two are ‘fair’ treaties for the tower. \par

Reviewer 4: ```suggestion
When the agent is not in critical condition, it is able to propose other treaties depending on their morality. The first treaty is detrimental for the tower and the other two are ‘fair’ treaties for the tower. \par
```

\begin{enumerate}

>     \item If the agent is very immoral, it will send a treaty that is detrimental for anyone above them. It is expected for this treaty to be rejected by any sensible agent’s and is done as a way of mimicking frustration or the desire to hurt others when morality is low.

Reviewer: No comma on agent's

Reviewer 3: ```suggestion
    \item If the agent is very immoral, it will send a treaty that is detrimental for anyone above them. It is expected for this treaty to be rejected by any sensible agents and is done as a way of mimicking frustration or the desire to hurt others when morality is low.
```

    \item If HP is below 60 leave at least 60\% of the food on the platform for 5 days. This treaty tries to encourage immoral agents in higher positions to share food. 
    \item If HP is below 10 leave at least 95\% of the food on the platform for 10 days. This treaty is aimed at moral agents that would be willing to start eating less until they are close to death.
\end{enumerate}

> Both of the last treaties are proposed randomly by the agent, without a specific proclivity for any of the two. The randomness of the treaty selection is done to increase the volitional behaviour of the agent. Characteristics that the agent inherently lacks if its personality is solely based on the TRA.

Reviewer: if its changed to in its

\subsection{Handling a Proposed Treaty}

> Following our strategy, agents that are in a bad mood will not accept treaties that they need to honour when they are in very low health, since they have low confidence in the capabilities of other agents. However, when in a good mood agents will accept treaties that must be honoured even at low health. Morality is an indicator of the agent’s desire to perform actions that will benefit others, as opposed to just benefiting the agent itself. Therefore, very moral agents will accept treaties that involve them consuming just enough food to survive, but not too little food that they die or too much that they are too greedy. Similarly, very immoral agents will only accept treaties where they consume more than enough food since they disregard the food needs of the other agents. The table below displays the mood and morality values that agents must satisfy to accept treaties with any given condition depending on the agent’s health, and the food the agent would have to leave according to the request of the treaty. \par

Reviewer: Following our strategy unclear. Change to something like In our strategy.

        mood $>$ 20             & Healthy (2)                           & m$>$20, 0$<$M$<$40        & m$>$20, 20$<$M$<$60       & m$>$20, 40$<$M$<$80       & m$>$20, 60$<$M$<$100      & m$>$20, 80$<$M        & N/A since 100$<$M     \\ \hline
        mood $>$ 40             & Average (3)                           & m$>$40, 0$<$M$<$40        & m$>$40, 20$<$M$<$60       & m$>$40, 40$<$M$<$80       & m$>$40, 60$<$M$<$100      & m$>$40, 80$<$M        & N/A since 100$<$M     \\ \hline
        mood $>$ 60             & Weak (4)                              & m$>$60, 0$<$M$<$40        & m$>$60, 20$<$M$<$60       & m$>$60, 40$<$M$<$80       & m$>$60, 60$<$M$<$100      & m$>$60, 80$<$M        & N/A since 100$<$M     \\ \hline
        mood $>$ 80             & SurvivalLevel (5)                     & m$>$80, 0$<$M$<$40        & m$>$80, 20$<$M$<$60       & m$>$80, 40$<$M$<$80       & m$>$80, 60$<$M$<$100      & m$>$80, 80$<$M        & N/A since 100$<$M     \\ \hline
    \end{tabular}}
\end{table}

> Treaties also have a specified duration and signature count, which are visible to any agent that wishes to sign the treaty. These are likely to affect an agent’s decision to accept a treaty and have been taken into account. If a treaty lasts longer than twice the reshuffle period, an agent deems it too far into the future to be able to predict what will happen and plans to reject the treaty. However, if a treaty has more than five signatures, we deem it to be a popular treaty. If the agent plans to reject the treaty as a result of any previous logic, the influence of peer pressure will cause the agent to reconsider accepting the treaty. Peer pressure is based on the concept of crowd suggestibility coined by Park, R. E.  in 1904 in his “Masse und Publikum”, \cite{Solr-413687}. The stubbornness of an agent is used as the likelihood to give in to peer pressure, so more stubborn agents are less likely to reconsider their response, and more likely to stick with their original decision. \par

Reviewer: Remove comma before citation

\section{Food Intake}\label{sec:food_intake}
For an agent to choose what to eat on a given day, they first take into account treaties to which they have agreed. These treaties take the highest priority when deciding how much food to take from the platform. Therefore, when the agent in question has accepted a treaty the choice becomes dependent on the treaty conditions. \par 
For each treaty that has been accepted, the agent follows the following logic: 
\begin{enumerate}
    \item They start by checking whether the condition stated within the specific treaty is satisfied, in which case the agreed-upon amount of food to leave on the platform is calculated. 
    \item In the case where the treaty request is not to leave food on the platform but to inform the sender of the treaty of our agent's HP, a message containing our current HP is sent to the sender agent. 
\end{enumerate}

> After all the treaty conditions have been dealt with the agent calculates how much food to eat itself. The calculation depends on the morality of the agent, how much food it has decided to leave on the platform and how much food it has decided to eat. It must be noted that during other processes external to the final decision of how much to eat the agent might have decided how much it would like to eat, a non-binding decision. \par

Reviewer: add i.e, before a non-binding decision

Reviewer 3: ```suggestion
After all the treaty conditions have been dealt with the agent calculates how much food to eat itself. The calculation depends on the morality of the agent, how much food it has decided to leave on the platform and how much food it has decided to eat. It must be noted that during other processes external to the final decision of how much to eat the agent might have decided how much it would like to eat, i.e a non-binding decision. \par
```

Reviewer 3: For future reviews @tomeaton17 , could u ask for changes like this directly so people can quickly commit them? Would be less work for everyone :D (done with ctrl+g I believe or in the insert a suggestion button)

Reviewer: Yes sorry, I didn't know about this feature.

\chapter{Team 3 Agent Design}\label{team_3_agent_design}

\section{The Agent}\label{sec:the_agent}
Team 3’s agents design is based on the Theory of Reasoned Action (TRA) and its extension, the Theory of Planned Behaviour (TPB). The agents are designed to have a shiftable personality dependent on the three variables in Figure \ref{fig:TRA-diagram} that TPB  \cite{AJZEN1991179} links to the behaviours of an agent. These three variables, together with a social network, define its personality as a whole, \cite{victor} and influence the actions it will take.
\begin{figure}[htb]
    \centering
    \includegraphics[width=0.5\linewidth]{005_team_3_agent_design/images/image1.jpg}

>     \caption{TRA explanation}

Reviewer 5: ```suggestion
    \caption{TRA explanation.}
```

    \label{fig:TRA-diagram}
\end{figure}

> The Theory of Reasoned Action was developed by Martin Fishbein in 1967 and was later improved by Icek Ajzen and Martin in 1980, with the Theory of Planned Behaviour. The TPB states that the behavioural intentions of individuals can be attributed to three interdependent factors: Perceived behaviour control, the subjective norm, and the attitude of the individual, as you can see in Figure \ref{fig:TRA-diagram}. It must be noted that the TPB is unable to accurately and consistently predict non-volitional behaviours of the individual. Therefore, TPB can only be applied to those actions for which a reason can be found. These are actions that occur as a reaction to others and that the individual has control over \cite{TRA}. This theory is applicable to agent decisions in the tower since the agent must make decisions in a reactionary manner, based on interactions with other agents.\par 

Reviewer 6: ```suggestion
The Theory of Reasoned Action was developed by Martin Fishbein in 1967 and was later improved by Icek Ajzen and Martin in 1980, with the Theory of Planned Behaviour. The TPB states that the behavioural intentions of individuals can be attributed to three interdependent factors: perceived behaviour control, the subjective norm, and the attitude of the individual, as you can see in Figure \ref{fig:TRA-diagram}. It must be noted that the TPB is unable to accurately and consistently predict non-volitional behaviours of the individual. Therefore, TPB can only be applied to those actions for which a reason can be found. These are actions that occur as a reaction to others and that the individual has control over \cite{TRA}. This theory is applicable to agent decisions in the tower since the agent must make decisions in a reactionary manner, based on interactions with other agents.\par 
```

> In order to apply the TPB to our agents, we define three agent variables, each corresponding to one of the three components of the theory. Using this approach, we have defined the agent variables to be: 

Reviewer 6: ```suggestion
In order to apply the TPB to the agents, we define three agent variables, each corresponding to one of the three components of the theory. Using this approach, we have defined the agent variables to be: 
```

\begin{enumerate}
    \item \textbf{Stubbornness}: it addresses perceived behaviour control, which is the influence that external agents have on the decisions that are to be taken.
    \item \textbf{Morality}: it addresses the subjective norm, which is one's idea of what is socially perceived to be right or wrong.
    \item \textbf{Mood}: it defines the attitude towards the action to take.
\end{enumerate}
These variables are modified as an agent interacts, which leads to our agent having mutable behavioural intentions, and making decisions with different “intended outcomes” as the variables change. \par

> In literature surrounding the substantive uniqueness of human behaviour \cite{TRA} Martin Fishbein argues that no two agents will react identically to the same circumstances. In practice, reactions tend to be normally distributed, with a random aspect to them. To account for this, all possible changes to  internal agent variables are implemented by specifying the range within which the variable can change. The final change to the variable is randomly selected from within this range. This characteristic of randomness in the agents’ behaviour compensates for the fact that the TPB does not take into account non-volitional behaviours, so including this randomness makes the agent act in a more human-like manner.\par 

Reviewer 6: ```suggestion
In literature surrounding the substantive uniqueness of human behaviour \cite{TRA} Martin Fishbein argues that no two agents will react identically to the same circumstances. In practice, reactions tend to be normally distributed, with a random aspect to them. To account for this, all possible changes to internal agent variables are implemented by specifying the range within which the variable can change. The final change to the variable is randomly selected from within this range. This characteristic of randomness in the agents’ behaviour compensates for the fact that the TPB does not take into account non-volitional behaviours, so including this randomness makes the agent act in a more human-like manner.\par 
```

> The next step after identifying our agent strategy was to think about the scenario and how the problem should be solved. Our first approach used classical game theory, which states that under normal circumstances, an unbiased agent would always prioritise themselves and their needs over anything else. Hence, this dictates that each agent should not act in a moral way at all. In our scenario, this leads to agents taking very large amounts of food, causing agents below them to have access to less food, and have a greater likelihood of death. However, we found research demonstrating how humans in a social environment tend to make decisions benefiting the collective \cite{batson_batson_todd_brummett_shaw_aldeguer_1995}. This contradicted traditional game theory views and made this approach unreasonable when trying to recreate a human type behaviour, following the two most common explanations of why humans violate classic game theory: enlightened self-interest and group identity.\par 

Reviewer 6: ```suggestion
The next step after identifying the agent strategy was to think about the scenario and how the problem should be solved. Our first approach used classical game theory, which states that under normal circumstances, an unbiased agent would always prioritise themselves and their needs over anything else. Hence, this dictates that each agent should not act in a moral way at all. In our scenario, this leads to agents taking very large amounts of food, causing agents below them to have access to less food, and have a greater likelihood of death. However, we found research demonstrating how humans in a social environment tend to make decisions benefiting the collective \cite{batson_batson_todd_brummett_shaw_aldeguer_1995}. This contradicted traditional game theory views and made this approach unreasonable when trying to recreate a human type behaviour, following the two most common explanations of why humans violate classic game theory: enlightened self-interest and group identity.\par 
```

\begin{enumerate}
    \item The floors the agent has been in before. This information will affect the agent’s mood after being reshuffled, as it would remember where they have been and know if the new floor is a good or bad one.  
    \item The last recorded HP value. Used to see changes to the agent's health.
    \item The agent’s friendship network. It is aware of the people it has met during their time in the tower. The agent’s friends are identified and are considered more or less close to the agent with an affection parameter that ranges from 0 to 1, with 0  representing a complete dislike for the other agent.
    \item The amount of food last eaten.
    \item The amount of food last seen on the platform.
    \item A moving average of food consumed.
    \item The age of the agent.
    \item The treaty we have proposed until it is accepted or rejected.
    \item The estimation of the reshuffle period.
    \item The HP our neighbour above last told us it had.
    \item The HP our neighbour below last told us it had.
\end{enumerate}

\subsection{Agent Generation}

> The defining variables of our agents are randomised when it is generated in order to avoid preconditioned agents and mimic different personalities. All three variables are defined to be in the range 0 to 100. Morality and mood are randomly allocated to a value in the range, but stubbornness is limited to a number between 0 and 75. As explained in our agent parameters, stubbornness represents the effect of external influences on an agent. A stubbornness of 100 would mean this agent is completely unwilling to listen and would ignore all incoming messages, effectively making them an agent that disrupts the tower. In order for this to not be the case, we added the artificial upper bound to the value of stubbornness.\par 

Reviewer 6: ```suggestion
The defining variables of the agents are randomised when it is generated in order to avoid preconditioned agents and mimic different personalities. All three variables are defined to be in the range 0 to 100. Morality and mood are randomly allocated to a value in the range, but stubbornness is limited to a number between 0 and 75. As explained in our agent parameters, stubbornness represents the effect of external influences on an agent. A stubbornness of 100 would mean this agent is completely unwilling to listen and would ignore all incoming messages, effectively making them an agent that disrupts the tower. In order for this to not be the case, we added the artificial upper bound to the value of stubbornness.\par 
```

As Shao et al. have shown in “Beyond Moral Reasoning: A Review of Moral Identity Research and Its Implications for Business Ethics” \cite{shao_aquino_freeman_2008}, one’s moral action is most heavily influenced by their moral personality, as opposed to their moral reasoning, which empirically seems to show only a minor influence on the moral behaviour of an agent \cite{blasi_1983}. Hence, we used the variance and unpredictability to initialise our agents randomly on the moral scale. This insight also allowed us to witness a purely emergent strategy and nature as opposed to a pre-programmed inherent morality.\par 

> While the implementation of initial randomness alone allowed for interesting strategies to occur, we also wanted our agents to adapt their morality on the basis of their community and associations, as argued by Shao. This was made possible by having all actions an agent does affect its emotion variables. \par 

Reviewer 6: ```suggestion
While the implementation of initial randomness alone allowed for interesting strategies to occur, we also wanted the agents to adapt their morality on the basis of their community and associations, as argued by Shao. This was made possible by having all actions an agent does affect its emotion variables. \par 
```

\section{Agent Communication}\label{sec:agent_communication}

> Communication between agents is fundamental for the possibility of collaboration and self-organisation between agents. Crowd behaviour is considered to be generated by individuals, context-dependent and dynamic, as defended by the modern foundation of crowd behaviour \cite{0d9bc1ee81234780b2bb6ecd02762d56}. In this project, communication between agents is based on two concepts, treaties and direct messages, which are equally exploited by our agent and combined with the conceptualization of friendship levels. 

Reviewer 6: ```suggestion
Communication between agents is fundamental for the possibility of collaboration and self-organisation between agents. Crowd behaviour is considered to be generated by individuals, context-dependent and dynamic, as defended by the modern foundation of crowd behaviour \cite{0d9bc1ee81234780b2bb6ecd02762d56}. In this project, communication between agents is based on two concepts, treaties and direct messages, which are equally exploited by the Team 3 agent and combined with the conceptualization of friendship levels. 
```

\subsection{The Agent’s Social Network}

> The agent forms relationships with others surrounding it as it communicate with each other. The agent remembers everyone it has met and assigns a level of friendship to it depending on its interactions. By doing this, the agent forms a social network that will affect its decisions, something that the Theory of Planned Behaviour explores in the “subjective norm” aspect of the theory. In the TPB,  all three main variables are interdependent and for this reason, the agent will have a better or worse opinion of someone it has just met depending on their morality at that point. \par

Reviewer 6: ```suggestion
The agent forms relationships with others surrounding it as they communicate with each other. The agent remembers everyone it has met and assigns a level of friendship to them depending on their interactions. By doing this, the agent forms a social network that will affect its decisions, something that the Theory of Planned Behaviour explores in the “subjective norm” aspect of the theory. In the TPB,  all three main variables are interdependent and for this reason, the agent will have a better or worse opinion of someone it has just met depending on their morality at that point. \par
```

> The large importance which our agent places on friendship is not only a logical interpretation of the TPB. In fact, it is a notion which philosophers and legislators have repeatedly confirmed throughout the ages. For example, the Aristotelian concept of friendship defends it as “the finest way we exercise practical reasoning”. Quoting Aristotle: “Friendship seems to hold cities together, and lawgivers seem to care for it more than for justice … and when men are friends, they have no need for justice”, friendship is considered the best form of human moral action by The Nicomachean Ethics \cite{sokolowski}.Thus, it is no surprise that friendship is fundamental in the decision-making process of the agent, and also in its responses to communication initiated by other agents. \par 

Reviewer 6: ```suggestion
The large importance which the agent places on friendship is not only a logical interpretation of the TPB. In fact, it is a notion which philosophers and legislators have repeatedly confirmed throughout the ages. For example, the Aristotelian concept of friendship defends it as “the finest way we exercise practical reasoning”. Quoting Aristotle: “Friendship seems to hold cities together, and lawgivers seem to care for it more than for justice … and when men are friends, they have no need for justice”, friendship is considered the best form of human moral action by The Nicomachean Ethics \cite{sokolowski}. Thus, it is no surprise that friendship is fundamental in the decision-making process of the agent, and also in its responses to communication initiated by other agents. \par 
```

> To provide an alternative perspective on friendship, McCullough et al. investigate the nature of prosocial action and gratitude’s place in transactions \cite{doi:10.1111/j.1467-8721.2008.00590.x}. Our friendship function allows us to assign a level of gratitude to our agent in response to the prosocial action of the other agents. In this way, we could more effectively alert the agent to the prosocial nature of another’s actions. By utilising the friendship function to, in turn, behave more prosocially to those who have done so to us, we embedded a system of gratitude in our agents’ actions. This allows the agents to be more capable of self-organising.\par 

Reviewer 6: ```suggestion
To provide an alternative perspective on friendship, McCullough et al. investigate the nature of prosocial action and gratitude’s place in transactions \cite{doi:10.1111/j.1467-8721.2008.00590.x}. The friendship function allows the agent to assign a level of gratitude in response to the prosocial action of the other agents. In this way, the agent could more effectively alert themselves to the prosocial nature of another’s actions. By utilising the friendship function to, in turn, behave more prosocially to those who have done so to us, we embedded a system of gratitude in the agents’ actions. This allows the agents to be more capable of self-organising.\par 
```

> We further developed our response to actions taken by foreign agents by evaluating the selflessness of the action taken. That is to say, the more severe the sacrifice another agent makes, the more grateful we become. This is based on McCullough’s definition that one experiences gratitude when “they have benefited from someone’s costly, intentional, voluntary effort on their behalf”. \par 

Reviewer 6: ```suggestion
We further developed the agent's response to actions taken by foreign agents by evaluating the selflessness of the action taken. That is to say, the more severe the sacrifice another agent makes, the more grateful the agent becomes. This is based on McCullough’s definition that one experiences gratitude when “they have benefited from someone’s costly, intentional, voluntary effort on their behalf”. \par 
```

> Emotions allow us to respond to reality in a self-protective and self-enhancing way \cite{ekman_1992}, it was this response that we wanted our agent to embody so that emotions could govern the evolution of their actions, by either leading them on a self-protective path, if circumstances caused selfish agents to prevail, or a self-enhancing manner, as cooperation would beget further cooperation through our system of gratitude.\par

Reviewer 6: ```suggestion
Emotions allow us to respond to reality in a self-protective and self-enhancing way \cite{ekman_1992}, it was this response that we wanted the agent to embody so that emotions could govern the evolution of their actions, by either leading them on a self-protective path, if circumstances caused selfish agents to prevail, or a self-enhancing manner, as cooperation would beget further cooperation through the agents' system of gratitude.\par
```

> This is the basis of our friendship metric and implies our agent will react more positively to a friend and will increase the friendship of those with whom it has positive interactions.

Reviewer 6: ```suggestion
This is the basis of our friendship metric and implies the agent will react more positively to a friend and will increase the friendship of those with whom it has positive interactions.
```

\end{table}
\begin{table}[htb]
    \centering
    \begin{tabular}{@{}lllll@{}}
    \toprule
    HP inc last tick [nonfriend]    & Stubbornness      & Morality         & Mood             & Friendship     \\ \midrule
    ask Food Taken                  & 0                 & 0                & 0                & -1             \\
    ask HP                          & -1                & 1                & 1.5              & 1              \\
    ask Intended Food Intake        & 0                 & 0                & 0                & 0              \\
    Request Take Food               & 0                 & 0                & 0                & 0              \\
    Request Leave Food              & -1                & 1                & 1.5              & 1              \\ \bottomrule
    \end{tabular}
\end{table}
\begin{table}[htb]
    \centering
    \begin{tabular}{@{}lllll@{}}
    \toprule
    Sum total                       & Stubbornness      & Morality         & Mood             & Friendship     \\ \midrule
    Request Leave Food              & 0                 & 0                & 0                & 0              \\ \bottomrule
    \end{tabular}
\end{table}

> This zero-sum agent is how our agent is currently set, it represents the neutral agent. One who possesses no inherent emotional bias. This was done by balancing the positive and negative reactions to a friend, and those to a non-friend separately. We noticed the need for this when our agent would initially get stuck at the +ve and -ve limits of certain emotions. As future work, it would be exciting to see the emergent strategies which we hope to observe as we introduce biases. For instance, an agent whose net sum of all interactions leaves them with a stubbornness increase, coupled with an inherently moral agent, may create strategies yet thought of. 

Reviewer 6: ```suggestion
This zero-sum agent is how the agent is currently set, it represents the neutral agent. One who possesses no inherent emotional bias. This was done by balancing the positive and negative reactions to a friend, and those to a non-friend separately. We noticed the need for this when the agent would initially get stuck at the positive and negative limits of certain emotions. As future work, it would be exciting to see the emergent strategies which we hope to observe as we introduce biases. For instance, an agent whose net sum of all interactions leaves them with a stubbornness increase, coupled with an inherently moral agent, may create strategies yet thought of. 
```

    \item If HP is below 60 leave at least 60\% of the food on the platform for 5 days. This treaty tries to encourage immoral agents in higher positions to share food. 
    \item If HP is below 10 leave at least 95\% of the food on the platform for 10 days. This treaty is aimed at moral agents that would be willing to start eating less until it is close to death.
\end{enumerate}
Both of the last treaties are proposed randomly by the agent, without a specific proclivity for any of the two. The randomness of the treaty selection is done to increase the volitional behaviour of the agent. Characteristics that the agent inherently lacks if its personality is solely based on the TRA.

\subsection{Handling a Proposed Treaty}
In our strategy, we have decided to perform complex evaluation of food treaties. To evaluate how risky a treaty is, the two key metrics that are included in treaties will be used. These are the minimum agent condition at which the treaty is activated, and the maximum food that the treaty allows an agent to take.\par

> Agents that are in a bad mood will not accept treaties that it  needs to honour when it  is  in very low health, since it  has  low confidence in the capabilities of other agents. However, when in a good mood agents will accept treaties that must be honoured even at low health. Morality is an indicator of the agent’s desire to perform actions that will benefit others, as opposed to just benefiting the agent itself. Therefore, very moral agents will accept treaties that involve consuming just enough food to survive, but not too little food that it  dies or too much that it  is  too greedy. Similarly, very immoral agents will only accept treaties where it  consumes more than enough food since it  disregards the food needs of the other agents. The table below displays the mood and morality values that agents must satisfy to accept treaties with any given condition depending on the agent’s health, and the food the agent would have to leave according to the request of the treaty.

Reviewer 6: ```suggestion
Agents that are in a bad mood will not accept treaties that it needs to honour when it is in very low health, since it  has low confidence in the capabilities of other agents. However, when in a good mood agents will accept treaties that must be honoured even at low health. Morality is an indicator of the agent’s desire to perform actions that will benefit others, as opposed to just benefiting the agent itself. Therefore, very moral agents will accept treaties that involve consuming just enough food to survive, but not too little food that it dies or too much that it is too greedy. Similarly, very immoral agents will only accept treaties where it consumes more than enough food since it disregards the food needs of the other agents. The table below displays the mood and morality values that agents must satisfy to accept treaties with any given condition depending on the agent’s health, and the food the agent would have to leave according to the request of the treaty.
```

    \end{tabular}}
\end{table}
Treaties also have a specified duration and signature count, which are visible to any agent that wishes to sign the treaty. These are likely to affect an agent’s decision to accept a treaty and have been taken into account. If a treaty lasts longer than twice the reshuffle period, an agent deems it too far into the future to be able to predict what will happen and plans to reject the treaty. However, if a treaty has more than five signatures, we deem it to be a popular treaty. If the agent plans to reject the treaty as a result of any previous logic, the influence of peer pressure will cause the agent to reconsider accepting the treaty. Peer pressure is based on the concept of crowd suggestibility coined by Park, R. E.  in 1904 in his “Masse und Publikum”, \cite{Solr-413687}. The stubbornness of an agent is used as the likelihood to give in to peer pressure, so more stubborn agents are less likely to reconsider their response, and more likely to stick with their original decision.

\section{Food Intake}\label{sec:food_intake}
When taking food, an agent considers, in order of importance:
\begin{enumerate}

>     \item The treaties that it  has  agreed to. These must be adhered to if the agent fulfils the condition specified in the treaty, and define the amount of food that the agent must leave. 

Reviewer 6: ```suggestion
    \item The treaties that it has agreed to. These must be adhered to if the agent fulfils the condition specified in the treaty, and define the amount of food that the agent must leave. 
```

    \item The health level and morality of the agent. These are used to calculate the food the agent will eat, alongside any values which have been specified in treaties or the message pipeline.
\end{enumerate}
The agent calculates the amount of food it decides to take in the following manner: 
\begin{enumerate}
    \item If the agent is not at critical health, a target HP is calculated, which is designed to make an agent's HP converge to 65 if run repeatedly. The food required to reach the target HP level is then calculated, and scaled up if the agent is immoral, or down if the agent is moral.
    \item If the agent is at critical health, the agent aims to escape this state by targeting the weak level HP. The morality of the agent will dictate if it takes just the food it needs from the platform, or more, or maybe even less. However, on the last day before an agent’s death, it must desire to take at least survival food from the platform. This logic is expressed in the diagram \ref{fig:food-diagram}.
\end{enumerate}
\begin{figure}[htb]
    \centering
    \includegraphics[width=0.5\linewidth]{005_team_3_agent_design/images/diagram1.jpg}

>     \caption{Food calculation process}

Reviewer 5: ```suggestion
    \caption{Food calculation process.}
```

\chapter{Team 3 Agent Design}\label{team_3_agent_design}

\section{The Agent}\label{sec:the_agent}

> Team 3’s agents design is based on the Theory of Reasoned Action (TRA) and its extension, the Theory of Planned Behaviour (TPB). The agents are designed to have a shiftable personality dependent on the three variables in Figure \ref{fig:TRA-diagram} that TPB  \cite{AJZEN1991179} links to the behaviours of an agent. These three variables, together with a social network, define its personality as a whole, \cite{victor} and influence the actions it will take.

Reviewer 7: ```suggestion
Team 3’s agent design is based on the Theory of Reasoned Action (TRA) and its extension, the Theory of Planned Behaviour (TPB). The agents are designed to have a shiftable personality dependent on the three variables in Figure \ref{fig:TRA-diagram} that TPB  \cite{AJZEN1991179} links to the behaviours of an agent. These three variables, together with a social network, define its personality as a whole, \cite{victor} and influence the actions it will take.
```

\begin{figure}[htb]
    \centering
    \includegraphics[width=0.5\linewidth]{005_team_3_agent_design/images/image1.jpg}
    \caption{TRA explanation.}
    \label{fig:TRA-diagram}
\end{figure}

> The Theory of Reasoned Action was developed by Martin Fishbein in 1967 and was later improved by Icek Ajzen and Martin in 1980, with the Theory of Planned Behaviour. The TPB states that the behavioural intentions of individuals can be attributed to three interdependent factors: perceived behaviour control, the subjective norm, and the attitude of the individual, as you can see in Figure \ref{fig:TRA-diagram}. It must be noted that the TPB is unable to accurately and consistently predict non-volitional behaviours of the individual. Therefore, TPB can only be applied to those actions for which a reason can be found. These are actions that occur as a reaction to others and that the individual has control over \cite{TRA}. This theory is applicable to agent decisions in the tower since the agent must make decisions in a reactionary manner, based on interactions with other agents.\par 

Reviewer 7: ```suggestion
The Theory of Reasoned Action was developed by Martin Fishbein in 1967 and was later improved by Icek Ajzen and Martin in 1980, with the Theory of Planned Behaviour. The TPB states that the behavioural intentions of individuals can be attributed to three interdependent factors: perceived behaviour control, the subjective norm, and the attitude of the individual, as shown in Figure \ref{fig:TRA-diagram}. It must be noted that the TPB is unable to accurately and consistently predict non-volitional behaviours of the individual. Therefore, TPB can only be applied to those actions for which a reason can be found. These are actions that occur as a reaction to others and that the individual has control over \cite{TRA}. This theory is applicable to agent decisions in the tower since the agent must make decisions in a reactionary manner, based on interactions with other agents.\par 
```
